src/realm/views.cpp

```diff
@@ -190,9 +199,121 @@ bool SortDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
     return total_ordering ? i.index_in_view < j.index_in_view : 0;
 }
 
-void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& distinct)
+void DescriptorOrdering::append_sort(SortDescriptor sort)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        m_descriptors.emplace_back(std::move(sort));
+    }
+    else if (descriptor_is_sort(orig_length - 1)) {
+        // merge them together by appending internally
+        m_descriptors[orig_length - 1].merge_with(std::move(sort));
+    }
+    else { // last descriptor is distinct, just add it
+        m_descriptors.emplace_back(std::move(sort));
+    }
+}
+
+void DescriptorOrdering::append_distinct(DistinctDescriptor distinct)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        // add an empty sort descriptor first to keep ordering
+        m_descriptors.emplace_back(SortDescriptor());
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+    else if (descriptor_is_distinct(orig_length - 1)) {
+        // add the next distinct operation, separated by a placeholder sort
+        m_descriptors.emplace_back(SortDescriptor());
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+    else { // last descriptor is sort, just add it
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+}
+
+bool DescriptorOrdering::descriptor_is_sort(size_t index) const
+{
+    // The list is kept in alternating order of sort/distinct/sort/distinct...
+    return index % 2 == 0;
+}
+
+bool DescriptorOrdering::descriptor_is_distinct(size_t index) const
+{
+    return !descriptor_is_sort(index);
+}
+
+const SortDescriptor& DescriptorOrdering::operator[](size_t ndx) const
+{
+    if (REALM_UNLIKELY(ndx >= m_descriptors.size())) {
+        throw LogicError(LogicError::descriptor_out_of_range);
+    }
+    return m_descriptors[ndx];
+}
+
+bool DescriptorOrdering::will_apply_sort() const
+{
+    size_t num_descriptors = m_descriptors.size();
+    for (size_t desc_ndx = 0; desc_ndx < num_descriptors; ++desc_ndx) {
+        if (descriptor_is_sort(desc_ndx) && m_descriptors[desc_ndx]) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool DescriptorOrdering::will_apply_distinct() const
+{
+    size_t num_descriptors = m_descriptors.size();
+    for (size_t desc_ndx = 0; desc_ndx < num_descriptors; ++desc_ndx) {
+        if (descriptor_is_distinct(desc_ndx) && m_descriptors[desc_ndx]) {
```

Comment:
This could be expressed using `std::any_of` if we weren't giving special significance to the indexes.